pkg/server/credtemplate/builder.go

```diff
@@ -0,0 +1,493 @@
+		applyX509SVIDAttributes(tmpl, attributes)
+	}
+
+	return tmpl, nil
+}
+
+func (b *Builder) BuildAgentX509SVIDTemplate(ctx context.Context, params AgentX509SVIDParams) (*x509.Certificate, error) {
+	tmpl, err := b.buildX509SVIDTemplate(params.SPIFFEID, params.PublicKey, params.ParentChain, pkix.Name{}, b.config.AgentSVIDTTL)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cc := range b.config.CredentialComposers {
+		attributes, err := cc.ComposeAgentX509SVID(ctx, params.SPIFFEID, params.PublicKey, x509SVIDAttributesFromTemplate(tmpl))
+		if err != nil {
+			return nil, err
+		}
+		applyX509SVIDAttributes(tmpl, attributes)
+	}
+
+	return tmpl, nil
+}
+
+func (b *Builder) BuildWorkloadX509SVIDTemplate(ctx context.Context, params WorkloadX509SVIDParams) (*x509.Certificate, error) {
+	subject := b.config.X509SVIDSubject
+	if params.Subject.String() != "" {
+		subject = params.Subject
+	}
+
+	tmpl, err := b.buildX509SVIDTemplate(params.SPIFFEID, params.PublicKey, params.ParentChain, subject, params.TTL)
+	if err != nil {
+		return nil, err
+	}
+
+	// The first DNS name is also added as the CN by default. This happens
+	// even if the subject is provided explicitly in the params for backwards
+	// compatibility. Ideally we wouldn't do override the subject in this
+	// case. It is still overridable via the credential composers however.
+	if len(params.DNSNames) > 0 {
+		tmpl.Subject.CommonName = params.DNSNames[0]
+		tmpl.DNSNames = params.DNSNames
+	}
+
+	for _, cc := range b.config.CredentialComposers {
+		attributes, err := cc.ComposeWorkloadX509SVID(ctx, params.SPIFFEID, params.PublicKey, x509SVIDAttributesFromTemplate(tmpl))
+		if err != nil {
+			return nil, err
+		}
+		applyX509SVIDAttributes(tmpl, attributes)
+	}
+
+	return tmpl, nil
+}
+
+func (b *Builder) BuildWorkloadJWTSVIDClaims(ctx context.Context, params WorkloadJWTSVIDParams) (map[string]interface{}, error) {
+	params.Audience = dropEmptyValues(params.Audience)
+
+	if params.SPIFFEID.IsZero() {
+		return nil, fmt.Errorf("invalid JWT-SVID ID: cannot be empty")
+	}
+	if err := api.VerifyTrustDomainMemberID(b.config.TrustDomain, params.SPIFFEID); err != nil {
+		return nil, fmt.Errorf("invalid JWT-SVID ID: %w", err)
+	}
+	if len(params.Audience) == 0 {
+		return nil, fmt.Errorf("invalid JWT-SVID audience: cannot be empty")
+	}
+
+	now := b.config.Clock.Now()
+
+	ttl := params.TTL
+	if ttl <= 0 {
+		ttl = b.config.JWTSVIDTTL
+	}
+	_, expiresAt := computeCappedLifetime(b.config.Clock, ttl, params.ExpirationCap)
+
+	attributes := credentialcomposer.JWTSVIDAttributes{
+		Claims: map[string]interface{}{
+			"sub": params.SPIFFEID.String(),
+			"exp": jwt.NewNumericDate(expiresAt),
+			"aud": params.Audience,
+			"iat": jwt.NewNumericDate(now),
+		},
+	}
+	if b.config.JWTIssuer != "" {
+		attributes.Claims["iss"] = b.config.JWTIssuer
+	}
+
+	for _, cc := range b.config.CredentialComposers {
+		var err error
+		attributes, err = cc.ComposeWorkloadJWTSVID(ctx, params.SPIFFEID, attributes)
```

Comment:
Should we detect and log a warning if a CredentialComposer plugin overwrites an attributes that was set by other CredentialComposer?

Reply:
I don't know. I could imagine situations where that might be useful, but they are contrived. Honestly, I don't know how people are going to use these plugins in the wild.

Reply:
I also agree that's hard to tell until we gather information about how this is used.